src/coreclr/vm/classlayoutinfo.cpp

```diff
@@ -295,6 +117,247 @@ namespace
+    {
+        ULONG cInstanceFields = 0;
+        BYTE alignmentRequirement = 0;
+        for (ULONG i = 0; i < cFields; i++)
+        {
+            FieldDesc* pField = &pFields[i];
+            if (pField->IsStatic())
+                continue;
+
+            cInstanceFields++;
+            CorElementType corElemType = pField->GetFieldType();
+            TypeHandle typeHandleMaybe{};
+
+            if (corElemType == ELEMENT_TYPE_VALUETYPE)
+            {
+                typeHandleMaybe = pByValueClassCache[i];
+
+                corElemType = typeHandleMaybe.AsMethodTable()->GetInternalCorElementType();
+                if (corElemType != ELEMENT_TYPE_VALUETYPE)
+                    typeHandleMaybe = TypeHandle();
+            }
+
+            pInfoArray[i].m_MD = pField->GetMemberDef();
+            pInfoArray[i].m_placement = GetFieldPlacementInfo(corElemType, typeHandleMaybe);
+
+            BYTE fieldAlignmentRequirement = (BYTE)pInfoArray[i].m_placement.m_alignment;
+
+            fieldAlignmentRequirement = min(fieldAlignmentRequirement, packingSize);
+
+            alignmentRequirement = max(alignmentRequirement, fieldAlignmentRequirement);
+
+            switch (fieldAlignmentRequirement)
+            {
+            case 1:
+            case 2:
+            case 4:
+            case 8:
+            case 16:
+            case 32:
+            case 64:
+                break;
+            default:
+                COMPlusThrowHR(COR_E_INVALIDPROGRAM, BFA_METADATA_CORRUPT);
+            }
+        }
+
+        *pNumInstanceFields = cInstanceFields;
+        *pAlignmentRequirement = alignmentRequirement;
+    }
+
+    void SetFieldOffsets(FieldDesc* pFields, ULONG cFields, LayoutRawFieldInfo* pInfoArray, ULONG cInstanceFields)
+    {
+        for (ULONG i = 0, iInstanceFieldInfo = 0; i < cFields; i++)
+        {
+            FieldDesc* pField = &pFields[i];
+            if (pField->IsStatic())
+                continue;
+
+            // We should only be placing unplaced fields at this point.
+            _ASSERTE(pField->GetOffset() == FIELD_OFFSET_UNPLACED
+                || pField->GetOffset() == FIELD_OFFSET_UNPLACED_GC_PTR
+                || pField->GetOffset() == FIELD_OFFSET_VALUE_CLASS);
+
+            _ASSERTE(iInstanceFieldInfo < cInstanceFields);
+            IfFailThrow(pField->SetOffset(pInfoArray[iInstanceFieldInfo++].m_placement.m_offset));
+        }
+    }
+
+    /// @brief Read the offsets for a type's fields from metadata for explicit layout.
+    /// @param pModule The module containing the type.
+    /// @param cl The metadata token of the type.
+    /// @param pFieldInfoArray The information about the instance fields of the type.
+    /// @param cInstanceFields The numer of instance fields in the type.
+    /// @param parentSize The size of the parent type's layout.
+    /// @return The end of the last field in this layout
```

Comment:
Is this doxygen? Do we have any other comments like this in runtime?

Reply:
Yeah this is the doxygen format. Got auto-inserted and was useful for setting things up. I see usages in Mono and the repo minipal, but not in coreclr. I can remove them if you want